TSI_INTERNAL_ERROR after upgrade to grpc 1.3.0

After upgrade to server python to grpcio 1.3.0 (and after regenerating proto files)
I have this:
`
E0505 20:13:50.851834103   32645 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
`
`
E0505 20:13:50.851868188   32645 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
`

and often

`
E0505 20:13:33.406171300   32645 chttp2_server.c:83]         Handshaking failed: {"created":"@1494015213.406134120","description":"Handshake read failed","file":"src/core/lib/security/transport/security_handshaker.c","file_line":239,"referenced_errors":[{"created":"@1494015213.406130326","description":"FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.c","file_line":456,"referenced_errors":[{"created":"@1494015213.406108664","description":"Handshake timed out","file":"src/core/lib/channel/handshaker.c","file_line":241}]}]}
`

The client code is Java (1.2.0 or 1.3.0). The certificate is generate by letsencrypt.

Client: java/android, grpc 1.3.0 (same results with grpc 1.2.0)
```
05-14 20:24:30.070 2033-2144/com.jobba.me D/ProfileProvider: [getProfile] error:io.grpc.StatusRuntimeException: UNAVAILABLE
05-14 20:24:30.070 2033-2033/com.jobba.me D/StartController: [lambda$doGetProfile$5] onError:io.grpc.StatusRuntimeException: UNAVAILABLE
```

the detail of the error is this:
```
io.grpc.StatusRuntimeException: UNAVAILABLE
    	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:227) ~[na:0.0]
    	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:208) ~[na:0.0]
    	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:141) ~[na:0.0]
    	at com.jobba.me.protocol.JobbaServiceGrpc$JobbaServiceBlockingStub.getProfile(JobbaServiceGrpc.java:1518) ~[na:0.0]
    	at com.jobba.me.providers.ProfileProvider.getProfile(ProfileProvider.java:127) ~[na:0.0]
    	at com.jobba.me.ui.controllers.StartController.lambda$doGetProfile$3(StartController.java:158) ~[na:0.0]
    	at com.jobba.me.ui.controllers.StartController$$Lambda$4.call(Unknown Source) ~[na:0.0]
    	at rx.internal.operators.SingleFromCallable.call(SingleFromCallable.java:41) ~[na:0.0]
    	at rx.internal.operators.SingleFromCallable.call(SingleFromCallable.java:29) ~[na:0.0]
    	at rx.Single.subscribe(Single.java:1967) ~[na:0.0]
    	at rx.Single$13$1.call(Single.java:2039) ~[na:0.0]
    	at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230) ~[na:0.0]
    	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55) ~[na:0.0]
    	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423) ~[na:0.0]
    	at java.util.concurrent.FutureTask.run(FutureTask.java:237) ~[na:0.0]
    	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:269) ~[na:0.0]
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) ~[na:0.0]
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) ~[na:0.0]
    	at java.lang.Thread.run(Thread.java:818) ~[na:0.0]
    Caused by: java.io.IOException: PROTOCOL_ERROR invalid settings id: -509
    	at io.grpc.okhttp.internal.framed.Http2.ioException(Http2.java:589) ~[na:0.0]
    	at io.grpc.okhttp.internal.framed.Http2.access$200(Http2.java:47) ~[na:0.0]
    	at io.grpc.okhttp.internal.framed.Http2$Reader.readSettings(Http2.java:304) ~[na:0.0]
    	at io.grpc.okhttp.internal.framed.Http2$Reader.nextFrame(Http2.java:162) ~[na:00]
    	at io.grpc.okhttp.OkHttpClientTransport$ClientFrameHandler.run(OkHttpClientTransport.java:868) ~[na:0.0]
    	... 3 common frames omitted
```

Server: python 2.7, grpcio 1.3.0
```
E0514 18:25:52.086246800   31522 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
E0514 18:25:52.086278226   31522 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E0514 18:25:53.255994191   31522 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
E0514 18:25:53.256027474   31522 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E0514 18:25:54.754042667   31522 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
E0514 18:25:54.754073034   31522 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E0514 18:25:57.276323568   31522 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
E0514 18:25:57.276358200   31522 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E0514 18:26:02.134398336   31522 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
E0514 18:26:02.134430648   31522 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E0514 18:26:09.914237780   31522 ssl_transport_security.c:429] SSL_read returned 0 unexpectedly.
E0514 18:26:09.914269613   31522 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
```

--- a/src/core/tsi/ssl_transport_security.c
+++ b/src/core/tsi/ssl_transport_security.c
@@ -411,15 +411,11 @@ static tsi_result do_ssl_read(SSL *ssl, unsigned char *unprotected_bytes,
   GPR_ASSERT(*unprotected_bytes_size <= INT_MAX);
   read_from_ssl =
       SSL_read(ssl, unprotected_bytes, (int)*unprotected_bytes_size);
-  if (read_from_ssl == 0) {
-    gpr_log(GPR_ERROR, "SSL_read returned 0 unexpectedly.");
-    return TSI_INTERNAL_ERROR;
-  }
-  if (read_from_ssl < 0) {
+  if (read_from_ssl <= 0) {
     read_from_ssl = SSL_get_error(ssl, read_from_ssl);
     switch (read_from_ssl) {
-      case SSL_ERROR_WANT_READ:
-        /* We need more data to finish the frame. */
+      case SSL_ERROR_ZERO_RETURN: /* Received a close_notify alert. */
+      case SSL_ERROR_WANT_READ:   /* We need more data to finish the frame. */
         *unprotected_bytes_size = 0;
         return TSI_OK;
       case SSL_ERROR_WANT_WRITE:

